### PR TITLE
[MRG+1] Support for returning deferreds in middlewares

### PR DESCRIPTION
**Pull request by ArturGaspar**
Adds support for returning deferreds in middlewares, and makes use of this to fix a limitation in RobotsTxtMiddleware.
#1471

### CODE REVIEW

File: scrapy/downloadermiddlewares/robotstxt.py [hunk 1]
```diff
@@ -52,9 +58,19 @@ def robot_parser(self, request, spider):
                 meta={'dont_obey_robotstxt': True}
             )
             dfd = self.crawler.engine.download(robotsreq, spider)
-            dfd.addCallback(self._parse_robots)
+            dfd.addCallback(self._parse_robots, netloc)
```
Comment (dangra) [inline scrapy/downloadermiddlewares/robotstxt.py]: :+1:

**Comment by dangra** [comment]
LGTM :clap:s

**Comment by dangra** [comment]
the patch looks amazing, if it can get 100% coverage on robotstxt middleware it would be awesome :)

https://codecov.io/github/scrapy/scrapy/scrapy/downloadermiddlewares/robotstxt.py?ref=f34d8e309e723e5cc54a892d0c2e58ccd6e01a23

**Comment by dangra** [comment]
@ArturGaspar thanks for increasing coverage of robotstxt mw, looks like it didn't reach 100% because of a missing branch check at https://codecov.io/github/scrapy/scrapy/scrapy/downloadermiddlewares/robotstxt.py?ref=af2acd5046c9349b4f1082b401bf8a8a9f9d93b9#l-76.

I saw you added deferred support to spidermiddlewares but please, keep this PR focused on adding support to downloader middleware, It is acceptable to port robotstxt becaues it is a good example on how to use this feature.

Spider Middleware have subtles differences like been able to return iterables so I prefer if its support is added in another PR.

**Comment by ArturGaspar** [comment]
How do I remove the commit for the spider middleware changes from the pull request?
Will I have to close it and make another one? Or should I revert the changes in another commit?

**Comment by dangra** [comment]
> How do I remove the commit for the spider middleware changes from the pull request?
> Will I have to close it and make another one? Or should I revert the changes in another commit?

use `git rebase --interactive` locally and remove the commit then `git push --force` to the same branch.

**Comment by ArturGaspar** [comment]
100% coverage for robotstxt.py now.

**Comment by dangra** [comment]
> 100% coverage for robotstxt.py now.
>  Hide all checks

cool! how much for 100% coverage in all lines affected by this patch? ;)

**Comment by ArturGaspar** [comment]
I'm not really sure this is the best way to test that part, but it needs to be covered so there it is.

**Comment by ArturGaspar** [comment]
It's a better idea indeed.

**Comment by dangra** [comment]
thanks!

**Comment by chekunkov** [comment]
yeah! finally someone got it done! :+1:

**Comment by pablohoffman** [comment]
Indeed! :clap:

**Pull request merged**